### ISSUE TITLE
STS 3.7.3 hangs on "populate auto detected configs"

### ISSUE THREAD

**Issue by kpage** [rzwitserloot/lombok#1127]
Hello, I installed lombok into STS 3.7.3.  I have added only the Apache Ivy and Gradle STS plugins.

Initially lombok worked well, but at some point STS became hung on the task "populate auto detected configs".  I had to kill the STS java process directly to exit.  I was unable to resolve this issue even after several restarts until I removed lombok from STS.ini, and then the problem went away.

Googling turned up this old issue related to lombok: https://issuetracker.springsource.com/browse/STS-3678

There was a link in a comment to this old lombok issue, but it seems the link no longer exists: https://code.google.com/p/projectlombok/issues/detail?id=645

I saw this in the error log.  Perhaps I have a separate issue where STS cannot contact some Spring server during the "populate auto detected configs" operation, but it seems like lombok somehow interferes with canceling this request:

eclipse.buildId=3.7.3.201602250940-RELEASE-e45
java.version=1.8.0_91
java.vendor=Oracle Corporation
BootLoader constants: OS=linux, ARCH=x86_64, WS=gtk, NL=en_US
Framework arguments:  -product org.springsource.sts.ide -clearPersistedState
Command-line arguments:  -os linux -ws gtk -arch x86_64 -product org.springsource.sts.ide -clearPersistedState

org.springframework.ide.eclipse.metadata
Warning
Thu Jun 09 10:26:25 ICT 2016
java.net.ConnectException: Connection timed out

java.lang.RuntimeException: java.net.ConnectException: Connection timed out
	at org.springframework.ide.eclipse.core.SpringCoreUtils.parseDocument(SpringCoreUtils.java:662)
	at org.springframework.ide.eclipse.core.SpringCoreUtils.parseDocument(SpringCoreUtils.java:639)
	at org.springframework.ide.eclipse.metadata.locate.DynamicWebProjectBeansConfigLocator.canLocateInProject(DynamicWebProjectBeansConfigLocator.java:245)
	at org.springframework.ide.eclipse.beans.core.model.locate.AbstractPathMatchingBeansConfigLocator.locateBeansConfigs(AbstractPathMatchingBeansConfigLocator.java:68)
	at org.springframework.ide.eclipse.beans.core.internal.model.BeansProject$2.run(BeansProject.java:1126)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.springframework.ide.eclipse.beans.core.internal.model.BeansProject.populateAutoDetectedConfigsAndConfigSetsInternally(BeansProject.java:1118)
	at org.springframework.ide.eclipse.beans.core.internal.model.BeansProject$1.run(BeansProject.java:1080)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
Caused by: java.net.ConnectException: Connection timed out
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at java.net.Socket.connect(Socket.java:538)
	at sun.net.NetworkClient.doConnect(NetworkClient.java:180)
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:432)
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:527)
	at sun.net.www.http.HttpClient.<init>(HttpClient.java:211)
	at sun.net.www.http.HttpClient.New(HttpClient.java:308)
	at sun.net.www.http.HttpClient.New(HttpClient.java:326)
	at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1169)
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1105)
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:999)
	at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:933)
	at sun.net.www.protocol.http.HttpURLConnection.followRedirect0(HttpURLConnection.java:2662)
	at sun.net.www.protocol.http.HttpURLConnection.followRedirect(HttpURLConnection.java:2584)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1770)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1441)
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.setupCurrentEntity(XMLEntityManager.java:647)
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startEntity(XMLEntityManager.java:1305)
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startDTDEntity(XMLEntityManager.java:1271)
	at com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.setInputSource(XMLDTDScannerImpl.java:263)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.dispatch(XMLDocumentScannerImpl.java:1167)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.next(XMLDocumentScannerImpl.java:1050)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:964)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:606)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:510)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:848)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:777)
	at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:243)
	at com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:339)
	at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:177)
	at org.springframework.ide.eclipse.core.SpringCoreUtils.parseDocument(SpringCoreUtils.java:656)
	... 8 more

**Comment by martinlippert** [comment]
I am not sure whether this is an issue with the Lombok installation or just plain STS, even though it disappeared after uninstalling the Lombok extension. Could you capture a few thread dumps while STS hangs? (using jps and jstack) Would be interesting to take a closer look. And if you have a small sample project that reproduces the issue would be also awesome.

**Comment by kpage** [comment]
Thanks for your reply, @martinlippert .  I tried to reproduce this again but was unable to.  I suspect it has nothing to do with lombok now, but that it was a coincidence due to having to restart STS to install/uninstall lombok.  If I can get it to happen again, I'll collect thread dumps and file with the STS issue tracker.

**Comment by martinlippert** [comment]
@kpage Ok, thanks for the feedback, and please open an issue when you come across this again. Thanks!!!

**Comment by kpage** [comment]
Just in case anyone hits on this issue in google search, I narrowed down the original "populate auto detected configs" issue to an old web.xml in a project I was opening infrequently.  The web.xml seems to be malformed with both an XML descriptor and a doctype:

```xml
<?xml version="1.0" encoding="ISO-8859-1"?>

<!DOCTYPE web-app PUBLIC
  "-//Sun Microsystems, Inc.//DTD Web Application 2.3//EN"         
  "http://java.sun.com/dtd/web-app_2_3.dtd">

<web-app>
   ...
```

I think STS was actually trying to request the url http://java.sun.com/dtd/web-app_2_3.dtd , which is not a real live url, and was just hanging on that.  It's still difficult to cancel that "populate auto detected configs" operation, but eventually the cancel did go through, so I guess STS is behaving correctly.  Once the web.xml was fixed by removing the doctype and setting namespaces on web-app instead, the issue went away.
